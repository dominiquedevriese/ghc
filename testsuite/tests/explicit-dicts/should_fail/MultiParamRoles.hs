{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DuplicateRecordFields #-}
{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE RoleAnnotations #-}
module MultiParamRoles where

import GHC.Base (Ord(..))

reverseOrd :: Ord a => Ord.Dict a
reverseOrd = undefined

class Ord a => Convert a b where
  convert :: a -> b

data Set a
type role Set nominal

insert :: Ord a => a -> Set a -> Set a
insert = undefined


f :: Convert a b => a -> Set a -> Set a
f = insert

g :: Ord a => a -> Set a -> Set a
g = f ((dict))
  where
    dict = Convert.Dict { parent1 = reverseOrd, convert = undefined }
