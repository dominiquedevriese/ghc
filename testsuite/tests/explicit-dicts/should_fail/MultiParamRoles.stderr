
MultiParamRoles.hs:26:5: error:
    • Explicit dictionary application to: ‘Convert a b’
      is not allowed because of instance incoherence:
      In the function type ‘a -> Set a -> Set a’
        Type variable ‘a’ has role Nominal
    • In the expression: f ((dict))
      In an equation for ‘g’:
          g = f ((dict))
            where
                dict = Convert.Dict {parent1 = reverseOrd, convert = undefined}
